### PR TITLE
update use of 'lsel' to match HEIF (#180)

### LINKED ISSUE #180: Implement new wording for 'lsel' and progressive decoding
The conclusion of the last meeting was to change the current spec draft and add a special value for 'lsel' that indicates the old behaviour of it missing.

This issue tracks that work.

### PR DESCRIPTION
**Pull request by cconcolato**
closes #180

***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/pull/184.html" title="Last updated on Feb 16, 2022, 11:41 PM UTC (9557bee)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/184/a1e8bc1...9557bee.html" title="Last updated on Feb 16, 2022, 11:41 PM UTC (9557bee)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -182,16 +182,16 @@ This property should be marked as essential.
...
+The semantics of the 'ispe' property as defined in [[!HEIF]] apply. More specifically, for AV1 images, it corresponds to the [=FrameWidth=] and [=FrameHeight=] values as defined in [[!AV1]] for a frame that depends on the presence and content of the 'lsel' and [=OperatingPointSelectorProperty=] properties as follows:
 
-- In the absence of a 'lsel' property associated with the item:
+- In the absence of a 'lsel' property associated with the item, or if it is present and its 'layer_id' value is set to 0xFFFF:
     - If no [=OperatingPointSelectorProperty=] is associated with the item, the 'ispe' shall document the dimensions of the last frame decoded when processing the operating point whose index is 0.
 
     - If an [=OperatingPointSelectorProperty=] is associated with the item, the 'ispe' property shall document the dimension of last frame decoded when processing the corresponding operating point.
```
Comment (leo-barnes) [inline index.bs:190]: Not part of this change:

```suggestion
    - If an [=OperatingPointSelectorProperty=] is associated with the item, the 'ispe' property shall document the dimensions of the last frame decoded when processing the corresponding operating point.
```

File: index.bs [hunk 2]
```diff
@@ -182,16 +182,16 @@ This property should be marked as essential.
 
 <h4 id ="image-spatial-extents-property">Image Spatial Extents Property</h4>
 
-The semantics of the 'ispe' property as defined in [[!HEIF]] apply. More specifically, for AV1 images, it corresponds to the [=FrameWidth=] and [=FrameHeight=] values as defined in [[!AV1]] for a frame that depends on the presence of the 'lsel' and [=OperatingPointSelectorProperty=] properties as follows:
+The semantics of the 'ispe' property as defined in [[!HEIF]] apply. More specifically, for AV1 images, it corresponds to the [=FrameWidth=] and [=FrameHeight=] values as defined in [[!AV1]] for a frame that depends on the presence and content of the 'lsel' and [=OperatingPointSelectorProperty=] properties as follows:
```
Comment (leo-barnes) [inline index.bs:185]: This sentence feels kind of hard to read, but I'm not completely sure how to improve it. Maybe something like this?

```suggestion
The semantics of the 'ispe' property as defined in [[!HEIF]] apply. More specifically, for AV1 images, it corresponds to the [=FrameWidth=] and [=FrameHeight=] values as defined in [[!AV1]] for the frame selected by the presence and content of the 'lsel' and [=OperatingPointSelectorProperty=] properties as follows:
```

File: index.bs [hunk 3]
```diff
@@ -215,7 +215,7 @@ NOTE: Although the clean aperture property ('clap') defined in [[!HEIF]] is appl
...
+[[!AV1]] expects the renderer to display only one frame within the selected [=Operating Point=], which should be the highest spatial layer that is both within the [=Operating Point=] and present within the temporal unit, but [[!AV1]] leaves the option for other applications to set their own policy about which frames are output, as defined in the general output process. AVIF sets a different policy, and defines how the 'lsel' property (mandated by [[!HEIF]] for layered images) is used to control which layer is rendered. According to [[!HEIF]], the interpretation of the 'layer_id' field in the 'lsel' property is codec specific. In this specification, the value 0xFFFF is reserved for a special meaning. In an 'lsel' property is associated with an [=AV1 Image Item=] but its 'layer_id' value is set to 0xFFFF, the renderer is free to render either: only the output image of the highest spatial layer, or to render all output images of all the intermediate layers, resulting in a form of progressive decoding. If a 'lsel' property is associated with an [=AV1 Image Item=] and the value of 'layer_id' is not 0xFFFF, the renderer is expected to render only the output image for that layer.
```
Comment (leo-barnes) [inline index.bs:218]:
```suggestion
[[!AV1]] expects the renderer to display only one frame within the selected [=Operating Point=], which should be the highest spatial layer that is both within the [=Operating Point=] and present within the temporal unit, but [[!AV1]] leaves the option for other applications to set their own policy about which frames are output, as defined in the general output process. AVIF sets a different policy, and defines how the 'lsel' property (mandated by [[!HEIF]] for layered images) is used to control which layer is rendered. According to [[!HEIF]], the interpretation of the 'layer_id' field in the 'lsel' property is codec specific. In this specification, the value 0xFFFF is reserved for a special meaning. If an 'lsel' property is associated with an [=AV1 Image Item=] but its 'layer_id' value is set to 0xFFFF, the renderer is free to render either: only the output image of the highest spatial layer, or to render all output images of all the intermediate layers, resulting in a form of progressive decoding. If a 'lsel' property is associated with an [=AV1 Image Item=] and the value of 'layer_id' is not 0xFFFF, the renderer is expected to render only the output image for that layer.
```

Comment (leo-barnes) [reply index.bs:218]: The UI seems to have problems rendering the suggestion. Minor typo:
`In an 'lsel' property` => `If an 'lsel' property`

**Comment by leo-barnes** [comment]
Looks good!

File: index.bs [hunk 4]
```diff
@@ -215,7 +215,7 @@ NOTE: Although the clean aperture property ('clap') defined in [[!HEIF]] is appl
...
+[[!AV1]] expects the renderer to display only one frame within the selected [=Operating Point=], which should be the highest spatial layer that is both within the [=Operating Point=] and present within the temporal unit, but [[!AV1]] leaves the option for other applications to set their own policy about which frames are output, as defined in the general output process. AVIF sets a different policy, and defines how the 'lsel' property (mandated by [[!HEIF]] for layered images) is used to control which layer is rendered. According to [[!HEIF]], the interpretation of the 'layer_id' field in the 'lsel' property is codec specific. In this specification, the value 0xFFFF is reserved for a special meaning. If an 'lsel' property is associated with an [=AV1 Image Item=] but its 'layer_id' value is set to 0xFFFF, the renderer is free to render either: only the output image of the highest spatial layer, or to render all output images of all the intermediate layers, resulting in a form of progressive decoding. If a 'lsel' property is associated with an [=AV1 Image Item=] and the value of 'layer_id' is not 0xFFFF, the renderer is expected to render only the output image for that layer.
```
Comment (wantehchang) [inline index.bs:218]: 1. IMPORTANT: In the sentence

>..., the renderer is free to render either: only the output image of the highest spatial layer, or to render all output images of all the intermediate layers, resulting in a form of progressive decoding.

the term "intermediate layers" does not include the highest spatial layer. Therefore this sentence is not accurate. I think it should read:

>..., the renderer is free to render either: only the output image of the highest spatial layer, or to render the output images of all the intermediate layers and the highest spatial layer, resulting in a form of progressive decoding.

Also, I suggest removing the colon (:) after the word "either".

2. Nit: Please search for

> an 'lsel'

and

> a 'lsel'

in this file and use one form consistently. In the current draft, only the `a 'lsel'` form is used.

**Review by wantehchang — CHANGES_REQUESTED** [review]
LGTM. I suggest a couple of changes.

File: index.bs [hunk 5]
```diff
@@ -249,7 +249,7 @@ class OperatingPointSelectorProperty extends ItemProperty('a1op') {
 <dfn noexport>op_index</dfn> indicates the index of the operating point to be processed for this item. Its value shall be between 0 and [=operating_points_cnt_minus_1=].
 
 <h5 id ="layer-selector-property">Layer Selector Property</h5>
-The 'lsel' property defined in [[!HEIF]] may be associated with an [=AV1 Image Item=]. The 'layer_id' indicates the value of the [=spatial_id=] to render. The value shall be between 0 and 3. The corresponding spatial layer shall be present in the bitstream and shall produce an output frame. Other layers may be needed to decode the indicated layer.
+The 'lsel' property defined in [[!HEIF]] may be associated with an [=AV1 Image Item=]. The 'layer_id' indicates the value of the [=spatial_id=] to render. The value shall be between 0 and 3, or the special value 0xFFFF. When a value between 0 and 3 is used, the corresponding spatial layer shall be present in the bitstream and shall produce an output frame. Other layers may be needed to decode the indicated layer.
```
[the remark this answers is not in the document]
Comment (cconcolato) [reply index.bs]: done

File: index.bs (hunk 4 above)
Comment (cconcolato) [reply index.bs]: done

**Comment by cconcolato** [comment]
@wantehchang can you update your review?

**Review by wantehchang — APPROVED** [review]
LGTM. Thank you for writing this update, Cyril.
